### PR TITLE
add detail dataset config feature by extra config file

### PR DESCRIPTION
**Pull request by fur0ut0**
README: https://github.com/fur0ut0/sd-yascripts/blob/feature/dataset_config/config_README-ja.md

Solve #58 #130

There are several major changes:

* add config file feature by `--config_file` option
* add subset feature for combining multiple directories into dataset
  * also enable configuration per subset
* support multiple datasets
  * combine datasets into dataset group and use it as pseudo dataset
* add switch feature of inter-dataset shuffle by `--bucket_shuffle_across_dataset`
* change metadata scheme for supporting multiple datasets and multiple subsets

I have conserved backward compatibility of current DreamBooth directory handling, which uses subdirectory structure for class tokens and the number of dataset repeat.

NOTE: Sorry for a bit messy history. Maybe squash merge would be better option.

### CODE REVIEW

**Comment by fur0ut0** [comment]
Also, there is one big issue that error message about invalid config file is not user friendly enough. I'm considering how to fix this.

It would be nice if you can share some idea.

**Comment by fur0ut0** [comment]
> I agree that the error message is too simple. I'm not familiar to voluptuous, so I have no idea how to improve it. If there is a way to know where is wrong when an error occurs, it will be good...

I agree. I would like to let it show what option name causes an error.

> I am wondering that where the shuffling of the buckets is done.
> If the bucket indices are not shuffled, the members of each batch are fixed for each epoch. I don't think that is desirable.

`BaseDataset.set_current_epoch()` invokes bucket shuffling.

The index passed to `__getitem__()` will be random when we use `torch.utils.data.DataLoader` to enable inter-dataset shuffling. Thus, I have decided moving bucket shuffling feature from `__getitem__()` to `set_current_epoch()`.

> Caption/tag dropping in Textual Inversion training will cause unexpected results. Because if token string is not included in the caption, no weight will be trained. It will be better to show some warning message if caption/tag dropping options are specified in TI training.

Caption dropout is already disabled for textual Inversion. The arguments of `ConfigSanitizer()` controls which dataset feature will be enabled.

```python
def __init__(self, support_dreambooth: bool, support_finetuning: bool, support_dropout: bool) -> None:
```

When we passed `support_dropout=False`, voluptuous does not parse options about caption dropout (`caption_dropout_rate` etc.). It will cause an error when an user passes these options.

As a result, the parameters such as `caption_dropout_rate` passed to `BaseSubset` will be default values (`0.0`) in this case.

> I think it will be better that the metadata is same to previous if the dataset has only one subset for backward compatibility.

OK, I will preserve backward compatibility of metadata when there is single dataset.

**Comment by fur0ut0** [comment]
Update:

* conserve existing metadata entries when config file is not given
* use `json.dumps` for storing value of new metadata entry `ss_datasets`
* enable shuffle feature of data loader by default
  * delete `--bucket_shuffle_across_dataset` option from argparse

> Oh, I had missed that too. It looks good! However if caption/tag dropping option is specified in TI training, it will be helpful to warn the user that dropping does not work.

Let me clarify the point.

In textual inversion, current implementation grants dropout options as unknown options, and passing these options via config file causes an error.

Error example: `voluptuous.error.MultipleInvalid: extra keys not allowed @ data['general']['caption_dropout_rate']`

This behavior is made by mimicking argparse behavior, because argparse also grants dropout options as unknown options.

Error example: `train_textual_inversion.py: error: unrecognized arguments: --caption_dropout_rate=0.1`

Do you mean you think we should ignore options instead of causing an error, or mean we should print additional message in these special cases and still cause an error?

**Comment by bmaltais** [comment]
I like this. One question I would like clarification on. When both CLI parameters are passed along the TOML config... which take precedence? My personal preference if that if a TOML file is used it sould always have precedence over the CLI options... That way on can override CLI options by setting config for specific dataset. I hope this is how it works.

Great work @fur0ut0 . This has been a pet peeve of mine sice I used kohya_ss sf-scripts. Finally there will be a way to easilly configure various dataset options.

**Comment by fur0ut0** [comment]
@kohya-ss 

> I will take time to review and test :)

Thanks a lot! Please feel free to contact me if you have any other questions.

@bmaltais 

> When both CLI parameters are passed along the TOML config... which take precedence?

The config file options are always prioritized over CLI options.

README describes about this behavior, but only in Japanese. Sorry for not providing English README in the current state. I will update README in the near future.

**Comment by bmaltais** [comment]
Cool. I like that the old CLI is still available for training those simple one dataset models... And be able to use the config file for more complex ones. I think this missing feature was holding back some users from using kohya for large training

File: library/train_util.py [hunk 1]
```diff
@@ -609,28 +657,29 @@ def __getitem__(self, index):
...
+      subset = self.image_to_subset[image_key]
       loss_weights.append(self.prior_loss_weight if image_info.is_reg else 1.0)
 
       # image/latentsを処理する
       if image_info.latents is not None:
-        latents = image_info.latents if not self.flip_aug or random.random() < .5 else image_info.latents_flipped
+        latents = image_info.latents if not subset.flip_aug or random.random() < .5 else image_info.latents_flipped
         image = None
       elif image_info.latents_npz is not None:
         latents = self.load_latents_from_npz(image_info, self.flip_aug and random.random() >= .5)
```
[the remark this answers is not in the document]
Comment (fur0ut0) [reply library/train_util.py]: Yes, I forgot changing this one.

I have fixed: https://github.com/kohya-ss/sd-scripts/pull/227/commits/629ac27e852650e4d0173f2ec7e0250bb1f0aacb

File: library/train_util.py [hunk 2]
```diff
@@ -730,153 +779,168 @@ def read_caption(img_path):
...
     if use_npz_latents:
-      npz_any = False
-      npz_all = True
+      flip_aug_in_subset = False
+
       for image_info in self.image_data.values():
+        subset = self.image_to_subset[image_info.image_key]
+
         has_npz = image_info.latents_npz is not None
         npz_any = npz_any or has_npz
```
[the remark this answers is not in the document]
Comment (fur0ut0) [reply library/train_util.py:941]: Sorry, I accidentally deleted necessary define statements.

I have fixed: https://github.com/kohya-ss/sd-scripts/pull/227/commits/1208e7b8f7b803fe9fa23d0c1947e434cc903abd

File: config_README-ja.md [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+  class_tokens = 'human'
+  keep_tokens = 1
+
+# これは fine tuning 方式のデータセット
+[[datasets]]
+resolution = [768, 768]
+batch_size = 2
+
+  [[datasets.subsets]]
+  image_dir = 'C:\piyo'
```
[the remark this answers is not in the document]
Comment (fur0ut0) [reply config_README-ja.md:52]: Yes, It supposed to able to have `image_dir`.

Thus, fine tuning subset should have at least two options: `image_dir` and `metadata_file`.

**Comment by fur0ut0** [comment]
> This toml file seems to cause an error: `voluptuous.error.MultipleInvalid: extra keys not allowed @ data['datasets'][0]['subsets'][0]['metadata_file']`
> 
> Maybe I am doing something wrong, any ideas?
> 
> ```
> [general]
> 
> # これは fine tuning 方式のデータセット
> [[datasets]]
> resolution = [768, 512]
> 
>   [[datasets.subsets]]
>   # image_dir = 'D:\Work\SD\Diffusers-DB\data\ds_test\ft_train1\images'
>   metadata_file = 'D:\Work\SD\Diffusers-DB\data\ds_test\ft_train1\ft1_lat.json'
> 
>   [[datasets.subsets]]
>   metadata_file = 'D:\Work\SD\Diffusers-DB\data\ds_test\ft_train2\ft2_cln.json'
> ```

We need to provide `image_dir` for every subset.

However, this error message seems weird, right? This comes from error message of voluptuous only shows the first one when there are multiple errors.

In this case, it firstly tries to parse config as DreamBooth dataset. This time it causes extra key error because DreamBooth dataset does not support "metadata_file" option.

After that, it tries to parse config as fine tuning dataset. This time it causes required key error because there is no "image_dir" option.

I think adding another required config like `is_dreambooth` to dataset would improve this weird behavior because we can deterministically parse dataset config. What do you think?

**Comment by fur0ut0** [comment]
> I also wonder that `matadata_file` is required for fine tuning dataset, so it would be possible to parse as a fine tuning data set first. If there is no `metadata_file`, then it would be the DreamBooth dataset.

I'm afraid that simply swapping parsing order still causes the same issue because DreamBooth dataset also has distinct options, "is_reg" and "caption_extension".

However, checking whether `metadata_file` exists before parsing seems nice. I will try implementing deterministic dataset parsing feature.

**Comment by fur0ut0** [comment]
Ah, I have found `image_dir` is not required for `FineTuningDataset` now when metadata file has absolute path information.

I think `image_dir` should be handled as optional in fine tuning dataset while handled as required in DreamBooth dataset. I will also fix this.

**Comment by fur0ut0** [comment]
Update:

* checking `metadata_file` existence before dataset parsing
  * This improves the confusing error message we have seen
* update README

> I would like to confirm one thing, is my understanding correct that we cannot mix subsets of DreamBooth and fine tuning as a subsets of a certain dataset?

You are right. These two subset types cannot be mixed into single dataset.

This is mainly because I have no idea how to compensate the number of regularization images when there are also fine tuning subsets. If you come up with some way, these different types of subsets might be able to mix together.

I have added description about this topic to README.
